lib/puppet/environments.rb

```diff
@@ -476,6 +488,20 @@ def get_conf(name)
       @loader.get_conf(name)
     end
 
+    # This implementation ensures the environment is loaded for duration
+    # of the block and is never reloaded, even if it expires.
+    #
+    def with_guarded_environment(env, &block)
+      Thread.current[GUARD_KEY] = true
```

Comment:
Yeah I was thinking thread local was needed if multi-threading was enabled in puppetserver. But using thread local means an environment could be reloaded by one thread while the environment is “guarded” by another. It’s probably better to use a Hash that’s visible across threads. Also all of the Cached methods are synchronized, so I need to move this method to the Environment class (otherwise it will serialize catalog compilation)

Reply:
I reworked this so that the `with_guarded_environment` method is on the `Environment` instance. The method itself is not synchronized, so it's possible for multiple JRuby/puppetserver request threads to call `with_guarded_environment` concurrently [when multithreading is enabled](https://github.com/puppetlabs/puppetserver/blob/73728f814efa902c32c6cb0ef9f12c24bed7764e/dev/puppetserver.conf#L93) . The Environment will increment its guard counter using a `lock` to synchronize updates. As a result, the environment won't be evicted if it is guarded by at least one guard. Note on MRI, the [`lock.synchronize` method doesn't do any locking](https://github.com/puppetlabs/puppet/blob/09aa69bbfafb88d8ca660e4d07de56f8811fdff2/lib/puppet/concurrent/lock.rb#L8-L12).

Reply:
I reworked this to set the guard/ref counter on the cache entry.